lib/new_relic/agent/instrumentation/grpc/client/instrumentation.rb

```diff
@@ -13,26 +13,48 @@ module GRPC
         module Client
           include NewRelic::Agent::Instrumentation::GRPC::Helper
 
-          def issue_request_with_tracing(method, requests, marshal, unmarshal,
+          def issue_request_with_tracing(grpc_type, method, requests, marshal, unmarshal,
             deadline:, return_op:, parent:, credentials:, metadata:)
             return yield unless trace_with_newrelic?
 
             segment = request_segment(method)
             request_wrapper = NewRelic::Agent::Instrumentation::GRPC::Client::RequestWrapper.new(@host)
             segment.add_request_headers(request_wrapper)
             metadata.merge!(metadata, request_wrapper.instance_variable_get(:@newrelic_metadata))
+            grpc_message = nil
+            grpc_status = 0
 
             NewRelic::Agent.disable_all_tracing do
-              NewRelic::Agent::Tracer.capture_segment_error(segment) do
+              begin
                 yield
+              rescue => e
+                NewRelic::Agent.notice_error(e)
+                grpc_status, grpc_message = grpc_status_and_message_from_exception(e)
+                raise
               end
             end
           ensure
+            add_attributes(segment, grpc_message: grpc_message, grpc_status: grpc_status, grpc_type: grpc_type)
             segment.finish if segment
           end
 
           private
 
+          def add_attributes(segment, attributes_hash)
+            return unless segment
+
+            attributes_hash.each do |attr, value|
+              segment.add_agent_attribute(attr, value)
+            end
+            segment.record_agent_attributes = true
+          end
+
+          def grpc_status_and_message_from_exception(exception)
+            return unless exception.message =~ /^(\d+):(\w+)\./
+
+            [Regexp.last_match(1).to_i, Regexp.last_match(2)]
```

Comment:
If there's an exception, its message text will look like this:

> 4:DEADLINE_EXCEEDED. (some JSON blob here)

Look for the "number", "colon", "word" to determine the status and message.